### ISSUE TITLE
Debug mode: stack smashing in `mi_verify_padding` on Linux

### ISSUE THREAD

**Issue by ojura** [microsoft/mimalloc#782]
This bug has left me a bit desperate because I think it's the worst instance of a heisenbug that I have ever seen.

This is on  Ubuntu 22.04, x86-64,  5.19.0-46-generic #47~22.04.1-Ubuntu, 12th Gen Intel(R) Core(TM) i9-12900, gcc 12 from the Ubuntu 22.04 repo; observed with mimalloc 2.1.2 and 2.0.7.

`mi_verify_padding` is consistently  failing on one line in a Qt-based app that I'm working on, in the sense that the backtrace up to `mi_verify_padding` looks exactly identical every single time we crash. This happens in 1 out of 5-1 out of 10 app runs, when we call Qt image resize 231 times for 231 images:

![image](https://github.com/microsoft/mimalloc/assets/9062353/3579b621-33c6-41ce-b174-605db9c60302)

During `mi_verify_padding`, someone smashes the stack by changing the highest byte of the 64bit canary of `mi_verify_padding`. I hardcoded the canary to be `cafef00ddeadbabe` at the beginning of our app, and in this case the `ca` is consistently morphed into `4a` which then causes the stack canary check at the end of the function to fail.

The worst part is the nondeterminism and time sensitivity. I tried a billion debugging techniques, from GDB watchpoints, recording with mozilla rr or undo.io. I went as far as to disable ASLR and write a Linux kernel module which attaches hardware watchpoints to the process using `register_user_hw_breakpoint`. The canary address on the stack when we crash is always the same with ASLR disabled, so it's possible to hardcode a watchpoint. However, a watchpoint at that address gets around 500k write events during a typical run. 

Using two additional hardware breakpoints, I implemented a breakpoint-ladder to enable the canary watchpoint _only_ when entering `mi_verify_padding` in the crash-inducing image resizing code path. After this, an event log made by the kernel module handler for the `mi_verify_padding` canary watchpoint has 231 events, which perfectly corresponds to 231 processed images. 

In other words, I tightened the scope of the canary watchpoint as tight as possible. But the problem is, as soon as I have a watchpoint enabled and a perf event handler attached, no matter how lightweight (just saving the IP and the new value),  _THE DAMN THING DOES NOT CRASH ANYMORE_!

The only debugging technique which hasn't scared the heisenbug away was sprinkling some additional stack guard checks, like so:

```c++
unsigned long long l;
asm("mov -0x8(%%rbp), %0;"
        "sub    %%fs:0x28,%0"
        : "=r"(l));
 if (l) __stack_chk_fail();
```

I added some checks to `mi_page_decode_padding` as well (but still for the canary of the calling stack frame of `mi_verify_padding`, because that's the one which gets corrupted). But this didn't help me to scope in on where the corruption is happening, because it is totally random which canary check gets tripped. Sometimes it is even the one added by gcc when returning.

Here's how I modified the function. Sometimes I still get corruption detected at the return, sometimes after `mi_page_decode_padding2` (this is a copy of `mi_page_decode_padding`, just with checks like these added).

```c++
static bool mi_verify_padding(const mi_page_t* page, const mi_block_t* block, size_t* size, size_t* wrong) {
  size_t bsize;
  size_t delta;
  unsigned long long l;
    asm("mov -0x8(%%rbp), %0;"
        "sub    %%fs:0x28,%0"
        : "=r"(l));
    if (l) __stack_chk_fail();
  bool ok = mi_page_decode_padding2(page, block, &delta, &bsize);
  asm("mov -0x8(%%rbp), %0;"
      "sub    %%fs:0x28,%0" : "=r" (l));
  if (l) __stack_chk_fail();
  *size = *wrong = bsize;
  asm("mov -0x8(%%rbp), %0;"
      "sub    %%fs:0x28,%0" : "=r" (l));
  if (l) __stack_chk_fail();
  if (!ok) return false;
  mi_assert_internal(bsize >= delta);
  *size = bsize - delta;
  asm("mov -0x8(%%rbp), %0;"
      "sub    %%fs:0x28,%0" : "=r" (l));
  if (l) __stack_chk_fail();
  if (!mi_page_is_huge(page)) {
    uint8_t* fill = (uint8_t*)block + bsize - delta;
    const size_t maxpad = (delta > MI_MAX_ALIGN_SIZE ? MI_MAX_ALIGN_SIZE : delta); // check at most the first N padding bytes
    mi_track_mem_defined(fill, maxpad);
    for (size_t i = 0; i < maxpad; i++) {
      if (fill[i] != MI_DEBUG_PADDING) {
        *wrong = bsize - delta + i;
        asm("mov -0x8(%%rbp), %0;"
            "sub    %%fs:0x28,%0" : "=r" (l));
        if (l) __stack_chk_fail();
        ok = false;
        break;
      }
    }
    mi_track_mem_noaccess(fill, maxpad);
  }
  return ok;
}
```

The fact that this is impossible to catch if we attach a kernel write watchpoint handler is boggling. Is it such a tight race condition that I can disturb it from the kernel by just having a light perf handler which saves the instruction pointer?

I cannot reproduce this at all on my i7-8750H laptop, it seems like a faster desktop CPU is needed.

I am not claiming that this is a mimalloc bug (it could somehow be user error -- we have a bunch of threads and 3rd party proprietary hardware support blobs linked in; we make sure to link mimalloc first for proper override; most of the threads are idling when this image scaling step is performed so I don't see a possible culprit). I would welcome any suggestions on how to illuminate this.

**Comment by ojura** [comment]
I was using system-shipped Qt 5.15.3, but it's possible that this bug (marked fixed in 5.15.3) is still alive, so maybe it's on the Qt people:

https://bugreports.qt.io/browse/QTBUG-92188

**Comment by ojura** [comment]
Qt people confirmed this is their bug in `QImage::scaled`. Closing.
